src/lib/labeller.ts

```diff
@@ -0,0 +1,187 @@
+ * kind of step in a pipeline.
+ *
+ * Example usage:
+ * ```typescript
+ * import { humanReadableLabel } from '@/lib/labellers';
+ * const step: FilterStep: {
+ *  // ...
+ * }
+ * const label = humanReadableLabel(step);
+ * ```
+ */
+import { StepMatcher } from '@/lib/matcher';
+import * as S from '@/lib/steps';
+
+/**
+ * human-readable labels for aggregation functions.
+ */
+const AGGFUNCTION_LABELS = {
+  sum: 'sum',
+  avg: 'average',
+  count: 'count',
+  min: 'min',
+  max: 'max',
+};
+
+/**
+ * default value separator in human-readable labels for multivalued fields.
+ */
+const MULTIVALUE_SEP = ', ';
+
+/**
+ * Compute a human-readable label from a filter step condition.
+ *
+ * Introspect the condition's operator to make the label the most precise possible.
+ * @param condition the filter step condition
+ */
+function filterExpression(
+  condition: S.FilterSimpleCondition | S.FilterComboAnd | S.FilterComboOr,
+): string {
+  if (S.isFilterComboAnd(condition)) {
+    return condition.and.map(cond => filterExpression(cond)).join(' and ');
+  } else if (S.isFilterComboOr(condition)) {
+    return condition.or.map(cond => filterExpression(cond)).join(' or ');
+  } else {
+    switch (condition.operator) {
+      case 'eq':
+        return `"${condition.column}" = ${condition.value}`;
+      case 'ne':
+        return `"${condition.column}" ≠ ${condition.value}`;
+      case 'gt':
+        return `"${condition.column}" > ${condition.value}`;
+      case 'ge':
+        return `"${condition.column}" ≥ ${condition.value}`;
+      case 'lt':
+        return `"${condition.column}" < ${condition.value}`;
+      case 'le':
+        return `"${condition.column}" ≤ ${condition.value}`;
+      case 'in':
+        return `"${condition.column}" in (${condition.value.join(MULTIVALUE_SEP)})`;
+      case 'nin':
+      default:
+        // only for typescript to be happy and see we always have a return value
+        return `"${condition.column}" not in (${condition.value.join(MULTIVALUE_SEP)})`;
+    }
+  }
+}
+
+/**
+ * Return the input string in lowercase except for the first character.
+ *
+ * @param label the strin to capitalize
+ */
+function capitalize(label: string) {
+  return label[0].toUpperCase() + label.slice(1).toLocaleLowerCase();
+}
+
+/**
+ * The `Labeller` class provides a human-readable label for each step.
+ */
+class StepLabeller implements StepMatcher<string> {
+  aggregate(step: Readonly<S.AggregationStep>) {
+    const dimensions = step.on.join(MULTIVALUE_SEP);
+    if (step.aggregations.length === 1) {
+      const [aggstep] = step.aggregations;
+      return `${capitalize(AGGFUNCTION_LABELS[aggstep.aggfunction])} on "${
+        aggstep.column
+      }" grouped by "${dimensions}"`;
+    } else {
+      const columns = step.aggregations.map(aggstep => aggstep.column).join(MULTIVALUE_SEP);
+      return `Aggregate "${columns}" grouped by "${dimensions}"`;
```

Comment:
As a general comment for multiple columns, it would be nicer and more understandable to display "Column1", "Column2" rather than "Column1, Column2"
![image](https://user-images.githubusercontent.com/34442515/64415534-cf694d00-d095-11e9-912c-1879aecda1f8.png)